cabal-install/src/Distribution/Client/CmdPath.hs

```diff
@@ -0,0 +1,404 @@
+          (intercalate "|" $ map pathOutputFormatString [minBound .. maxBound])
+          pathOutputFormatParser
+          pathOutputFormatPrinter
+      )
+  , option
+      []
+      ["compiler-info"]
+      "Print information of the project compiler"
+      pathCompiler
+      (\v flags -> flags{pathCompiler = v})
+      (yesNoOpt showOrParseArgs)
+  ]
+    <> map pathOption [minBound .. maxBound]
+  where
+    pathOption s =
+      option
+        []
+        [pathName s]
+        ("Print cabal's " <> pathName s)
+        pathDirectories
+        (\v flags -> flags{pathDirectories = Flag $ concat (flagToList (pathDirectories flags) <> flagToList v)})
+        (noArg (Flag [s]))
+
+-- | A path that can be retrieved by the @cabal path@ command.
+data ConfigPath
+  = ConfigPathCacheDir
+  | ConfigPathLogsDir
+  | ConfigPathStoreDir
+  | ConfigPathConfigFile
+  | ConfigPathInstallDir
+  deriving (Eq, Ord, Show, Enum, Bounded)
+
+-- | The configuration name for this path.
+pathName :: ConfigPath -> String
+pathName ConfigPathCacheDir = "cache-dir"
+pathName ConfigPathLogsDir = "logs-dir"
+pathName ConfigPathStoreDir = "store-dir"
+pathName ConfigPathConfigFile = "config-file"
+pathName ConfigPathInstallDir = "installdir"
+
+-------------------------------------------------------------------------------
+-- Action
+-------------------------------------------------------------------------------
+
+-- | Entry point for the 'path' command.
+pathAction :: NixStyleFlags PathFlags -> [String] -> GlobalFlags -> IO ()
+pathAction flags@NixStyleFlags{extraFlags = pathFlags', ..} cliTargetStrings globalFlags = withContextAndSelectors AcceptNoTargets Nothing flags [] globalFlags OtherCommand $ \_ baseCtx _ -> do
+  let pathFlags =
+        if pathCompiler pathFlags' == NoFlag && pathDirectories pathFlags' == NoFlag
+          then -- if not a single key to query is given, query everything!
+
+            pathFlags'
+              { pathCompiler = Flag True
+              , pathDirectories = Flag [minBound .. maxBound]
+              }
+          else pathFlags'
+  when (NoFlag == pathOutputFormat pathFlags) $
+    dieWithException verbosity (CmdPathRequiresOutputFormat (map pathOutputFormatString [minBound .. maxBound]))
+  when (not $ null cliTargetStrings) $
+    dieWithException verbosity CmdPathAcceptsNoTargets
+  when (buildSettingDryRun (buildSettings baseCtx)) $
+    dieWithException verbosity CmdPathCommandDoesn'tSupportDryRun
+
+  compilerPathOutputs <-
+    if not $ fromFlagOrDefault False (pathCompiler pathFlags)
+      then pure Nothing
+      else do
+        (compiler, _, progDb) <- runRebuild (distProjectRootDirectory . distDirLayout $ baseCtx) $ configureCompiler verbosity (distDirLayout baseCtx) (projectConfig baseCtx)
+        compilerProg <- requireCompilerProg verbosity compiler
+        (configuredCompilerProg, _) <- requireProgram verbosity compilerProg progDb
+        pure $ Just $ mkCompilerInfo configuredCompilerProg compiler
```

Comment:
Honestly, I hate this, I much rather would have some kind of rule system like shake, so we could depend on certain targets, in this case project base context.